addons/binding/org.openhab.binding.globalcache/src/main/java/org/openhab/binding/globalcache/handler/GlobalCacheHandler.java

```diff
@@ -621,89 +644,119 @@ protected void connect() {
 
             // If device doesn't have a serial module, just open the command connection
             if (!deviceSupportsSerial1()) {
-                if (deviceConnect(commandConnection)) {
+                if (commandConnect(commandConnection)) {
                     markThingOnline();
                     deviceIsConnected = true;
                     return;
                 }
             } else {
                 // Open the command connection and either 1 or 2 serial connections
-                if (deviceConnect(commandConnection) && deviceConnect(serial1Connection)) {
+                if (commandConnect(commandConnection) && serialConnect(serial1Connection)) {
                     if (deviceSupportsSerial2()) {
-                        if (deviceConnect(serial2Connection)) {
+                        if (serialConnect(serial2Connection)) {
                             markThingOnline();
                             deviceIsConnected = true;
+                            startSerialReaders();
                             return;
                         }
                     } else {
                         markThingOnline();
                         deviceIsConnected = true;
+                        startSerialReaders();
                         return;
                     }
                 }
             }
             disconnect();
         }
 
-        private boolean deviceConnect(DeviceConnection conn) {
+        private boolean commandConnect(DeviceConnection conn) {
             logger.debug("Connecting to {} port for thing {} at IP {}", conn.getName(), thingID(), conn.getIP());
-
-            // open socket
+            if (!openSocket(conn)) {
+                return false;
+            }
+            // create streams
             try {
-                conn.setSocket(new Socket());
-                conn.getSocket().bind(new InetSocketAddress(ifAddress, 0));
-                conn.getSocket().connect(new InetSocketAddress(conn.getIP(), conn.getPort()), SOCKET_CONNECT_TIMEOUT);
+                conn.setCommandIn(new BufferedReader(new InputStreamReader(conn.getSocket().getInputStream())));
+                conn.setCommandOut(new DataOutputStream(conn.getSocket().getOutputStream()));
             } catch (IOException e) {
-                logger.debug("Error connecting to {} port for thing {} at IP {}, exception={}", conn.getName(),
+                logger.debug("Error getting streams to {} port for thing {} at {}, exception={}", conn.getName(),
                         thingID(), conn.getIP(), e.getMessage());
                 markThingOfflineWithError(ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, e.getMessage());
-                deviceDisconnect(conn);
+                commandDisconnect(conn);
```

Comment:
Maybe I'm missing something (and I would not rule that out), but I'm not seeing the recursion.  It would be a bit unusual to not get the streams immediately after successfully opening the socket.  Nonetheless, when this IOException happens, the socket will be closed (deviceDisconnect) and the connection will be marked as disconnected.  An IOException could occur when the socket is closed, and that will be caught and logged.  On the next connection attempt, if the condition that caused the IOException persists, the above scenario would occur again.  But, I don't see where the recursion occurs.

Having said that, I do see a separate issue that I'm going to look at that I think has been there since the initial release.